Add histogram class and related functionality

codejaeger opened the pull request:
<!-- Pull Requests MUST come from topic branch based on develop, and NEVER on `master) --->

### Description

This contains the initial implementation of **histogram** functionality for STL containers for GIL images. Basic tests checking for correctness of histogram have been provided. Usage syntax and distinction between compatible and incompatible containers have been illustrated in a example file.

There was some dicussion regarding the name of the function (which now is **image_histogram**) in the mailing thread ([link](https://lists.boost.org/boost-gil/2020/05/0433.php)). I chose on 'image_histogram' since the name 'histogram' is used as the class name in Boost.Histogram and might cause problems in namespace mix-up. Another option could be **histogram_1d** since these functions are meant for that purpose.
<!-- What does this pull request do? -->

### Dependency

Optional - Docs for the this implementation are running in a parallel PR #503.

### References

[Discussions](https://lists.boost.org/boost-gil/2020/05/0410.php) on boost-gil mailing threads.

<!-- Any links related to this PR: issues, other PRs, mailing list threads, StackOverflow questions, etc. -->

### Tasklist

<!-- Add YOUR OWN TASK(s), especially if your PR is a work in progress -->

- [x] Add test case(s)
- [x] Provide documentation
- [x] Ensure all CI builds pass
- [x] Review and approve

lpranam commented on include/boost/gil/histogram.hpp:
```diff
@@ -0,0 +1,247 @@
+>
+inline Container image_histogram(SrcView const& srcview)
+{
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    static_assert(std::is_arithmetic<typename Container::value_type>::value,
+                    "Improper container type for images.");
+    static_assert(std::is_unsigned<typename channel_type<SrcView>::type>::value,
+                    "Improper container type for signed images.");
+
+    using channel_t = typename channel_type<SrcView>::type;
+    using pixel_t = pixel<channel_t, gray_layout_t>;
+
+    Container hist(std::numeric_limits<channel_t>::max() + 1);
+    
+    for_each_pixel(color_converted_view<pixel_t>(srcview), [&hist](pixel_t const& p) {
+        ++hist[p];
+    });
```
this is code repetition you should consider calling `image_histogram(SrcView const& srcview, Container &hist)`

codejaeger replied:
Maybe make a image_histogram_impl(SrcView const& srcview, Container &hist) without the constraints on the template type is what you meant?

mloskot replied:
@lpranam 
> I have some objections with the code format(template parameters)

What are the objections exactly?

Generally, as it was agreed quite some time ago, we try to keep the lines up to [100 characters long](https://github.com/boostorg/gil/blob/develop/CONTRIBUTING.md#guidelines).
Second, parameters are often long (e.g. `mp11::mp_size<typename Layout::color_space_t>::value`) and it is often a good idea to keep names of template parameters self-descriptive, avoid cryptic abbreviation.

So, this for example would fit that limit:

```cpp
template <typename T>
struct is_indexable<T, decltype((void) ++declval<T>()[0] , 0)> : std::true_type {} ;
```

but this would not, hence each parameter goes into separate lines and `<` and `>` are formatted similarly to blocks delimited with `{` and `}` 

```cpp
template <typename ChannelValue, typename Layout>
struct pixel :
    detail::homogeneous_color_base
    <
        ChannelValue,
        Layout,
        mp11::mp_size<typename Layout::color_space_t>::value
    >
{
...
};
```

That makes the whole guideline very simple, without too many exceptions or sub-rules.

The only other, subjectively, readable formatting of highly templated source code that I know is one used by Boost.Spirit or other libraries by [Joel de Guzman](https://github.com/cycfi/), Michael Caisse et. al.

mloskot replied:
@lpranam If you don't have any proposal on better formatting fitting template parameters list in 100 characters long lines, then I'd suggest to mark this conversation as resolved.

codejaeger replied:
@mloskot Is it necessary to break 

`histogram<boost::mp11::mp_at<bin_t, boost::mp11::mp_size_t<Dimensions>>...> sub_histogram() {}`

into 

```
    histogram
    <   
        boost::mp11::mp_at
        <
            bin_t,
            boost::mp11::mp_size_t<Dimensions>
        >...
    > sub_histogram() 
    {}
```

The former fits within 100 characters.

mloskot replied:
@codejaeger No, if it fits in line long for 100 characters, then no need for breaking it.
In my earlier comments I just used the snippet based on your code as an example.

lpranam commented on include/boost/gil/histogram.hpp:
```diff
@@ -0,0 +1,247 @@
+    using pixel_t = pixel<channel_t, gray_layout_t>;
+    
+    for_each_pixel(color_converted_view<pixel_t>(srcview), [&hist](pixel_t const& p) {
+        if (p < hist.size()) ++hist[p];
```
instead shouldn't this create equal size bins and store the data according to the bins!?

codejaeger replied:
I was not sure how we could convey that piece of information to the end user i.e. specify what the bin size they should use when accessing the values.

lpranam commented on test/core/histogram/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+  unset(_name)
+  unset(_target)
+  unset(_test)
+endforeach()
```
missing endline

lpranam commented on test/core/histogram/Jamfile:
```diff
@@ -0,0 +1,11 @@
+
+import testing ;
+
+run histogram.cpp ;
```
missing endline

lpranam commented on example/histogram_impl.cpp:
```diff
@@ -0,0 +1,98 @@
+	// image_histogram(view(img),st);							  		// Does not Work
+
+	return 0;
+}
```
missing endline

lpranam commented on include/boost/gil/histogram.hpp:
```diff
@@ -0,0 +1,247 @@
+    using channel_t = typename channel_type<SrcView>::type;
+    using pixel_t = pixel<channel_t, gray_layout_t>;
+
+    Container hist(std::numeric_limits<channel_t>::max() + 1);
```
I don't think this is a good idea! We also have pixel types like RGB64!

codejaeger replied:
Could we put a check on the channel size for specific container types? It wouldn't be clean though.

mloskot replied:
Or have our own containers as @lpranam suggested in https://lists.boost.org/boost-gil/2020/06/0437.php

codejaeger commented on include/boost/gil/histogram.hpp:
```diff
@@ -0,0 +1,247 @@
+//
+
+#ifndef BOOST_GIL_EXTENSION_HISTOGRAM_HISTOGRAM_HPP
+#define BOOST_GIL_EXTENSION_HISTOGRAM_HISTOGRAM_HPP
```
Yes, the first is a typo. It was first in the extension directory I forgot to update it later. I was myself not sure of the correct directory for it. Lets wait for others' opinions.

codejaeger commented on include/boost/gil/histogram.hpp:
```diff
@@ -0,0 +1,247 @@
+namespace detail {
+
+// Used to check if container implements operator[] and operator++
+template <typename T, typename = int>
+struct is_indexable : std::false_type {} ;
+
+template <typename T>
+struct is_indexable
+        <
+            T,
+            decltype((void) ++declval<T>()[0] , 0)
+        > : std::true_type {} ;
+
+// Used to check if container is resizable i.e. implements resize() member function
+template <typename T, typename = int>
+struct is_resizable : std::false_type {} ;
+
+template <typename T>
+struct is_resizable
+        <
+            T,
+            decltype((void) declval<T>().resize(1), 0)
+        > : std::true_type {} ;
+
+// Used to check if container is key-value type and the key is of std::integral type
+template <typename T, typename = void>
+struct has_integral_key_type : std::false_type {} ;
+
+template <typename T>
+struct has_integral_key_type
+        <
+            T,
+            typename std::enable_if<std::is_integral<typename T::key_type >::value >::type 
+        > : std::true_type {} ;
+
+} // namespace detail
```
Should I add it to concepts/detail/type_traits.hpp?

codejaeger commented on include/boost/gil/histogram.hpp:
```diff
@@ -0,0 +1,247 @@
+                detail::is_indexable<Container>::value &&
+                detail::is_resizable<Container>::value, int>::type = 0
+>
+inline void image_histogram(SrcView const& srcview, Container &hist)
```
Will do.

codejaeger commented:
I have ideas I would like to share further on the current approach. 
I added a diagram to explain better

![histogram](https://user-images.githubusercontent.com/32168969/83555934-b01c2500-a52c-11ea-8203-942ad4296726.png)

This approach will surely ease extending the implementation design in the future to support both multi dimensional histograms . 
In brief, I chose this approach because 

- it will simplify the usage for the end user (just deal with the simple function call syntax)
- handle cases like when a call is made to histogram_2d with a specific container type but the image provided is gray scale so the underlying implementation should call histogram_1d_impl
- simplify providing extensions of other container types, at most they need to replicate the procedure if not use the already present ones

So basically the first overloads make sure appropriate type checks are performed on both image and container type. The second overload would choose to call the overload specific to the requirements and the third overload can be easily extended to support higher dimensional containers (of any type). 
I would like to know if this seems right or there are will be some implications of using this approach.

mloskot commented:
@codejaeger Generally, I think the idea of the `histogram` class is a fine one, it looks good. For someone experienced with legacy GIL code, shift from metafunctions to `constexpr` functions and their invocations may be surprising :-)

A few extra high-level suggestions:
1. Avoid `get_` and `set_` prefix, so `get_dimension()` should read `dimension()`
2. Prefix all predicates with `is_` consistently, so 
    - `tuple_size_compatible` -> `is_tuple_size_compatible`
    - `tuple_type_compatible` -> `is_tuple_type_compatible`
3. The `cast_to_histogram_key` should read `make_histogram_key`
4. The `test/core/histogram/histogram.cpp` should be split into multiple files, for example
    - `fill.cpp` with tests for variety of image types
    - `key.cpp` with tests for `key_from_*`
    - `access.cpp` with tests basic accessing of histogram entries
    - `sub_histogram.cpp` with tests specific to sub-s
    - Some `BOOST_TEST` checks should become `static_assert`, for example
         ```cpp
         static_assert(h1.get_dimension() == h2.get_dimension(), ""); // not BOOST_TEST
         ```
    - `dimension.cpp`, `is_*.cpp` with compile-time only tests (`compile` and `compile-fail` rule in `Jamfile`)
    - `helpers.cpp` with run-time tests of `is_*` predicates and others
5. You need to clean up mixing `std::size_t` with `std::ptrdiff_t` with `int`. 
    It is important to remember that in `view(x, y)` the `x` and `y` are `std::ptrdiff_t`. This will also avoid numerous compilation warnings.
6. I am also not clear about the purpose of the sub-histogram, so I'd like to learn more about the motivation behind it.

codejaeger commented on include/boost/gil/histogram.hpp:
```diff
@@ -0,0 +1,298 @@
+#ifndef BOOST_GIL_HISTOGRAM_HPP
+#define BOOST_GIL_HISTOGRAM_HPP
+
+#include <boost/gil/concepts/concept_check.hpp>
+#include <boost/gil/metafunctions.hpp>
+#include <boost/gil/pixel.hpp>
+#include <boost/container_hash/extensions.hpp>
+#include <boost/mp11.hpp>
+
+#include <type_traits>
+#include <unordered_map>
```
@mloskot should the **container_hash/extensions.hpp** be shifted? It is actually quite confusing because boost::hash_value is actually found in another place at boost/functional/hash.hpp. I will look into this and report back.

mloskot replied:
@codejaeger I meant something like this, which is groups, then folder-based and alphabetical order within groups:

```
#include <boost/gil/metafunctions.hpp>
#include <boost/gil/pixel.hpp>
#include <boost/gil/concepts/concept_check.hpp>

#include <boost/mp11.hpp>
#include <boost/container_hash/extensions.hpp>

#include <type_traits>
#include <unordered_map>
```

lpranam commented on test/core/histogram/fill.cpp:
```diff
@@ -0,0 +1,151 @@
+    7, 8, 7, 8, 7, 8, 7, 8
+};
+
+void fill_gray_image(std::ptrdiff_t width, std::ptrdiff_t height, std::uint8_t* matrix, gil::gray8_view_t const& view)
```
could use `interleaved_view` maybe?

lpranam commented on test/core/histogram/fill.cpp:
```diff
@@ -0,0 +1,151 @@
+    }
+}
+
+void fill_rgb_image(std::ptrdiff_t width, std::ptrdiff_t height, std::uint8_t* matrix, gil::rgb8_view_t const& view)
```
could use `interleaved_view` maybe?

lpranam commented on test/core/histogram/cumulative.cpp:
```diff
@@ -0,0 +1,75 @@
+    check_cumulative_histogram();
+
+    return boost::report_errors();
+}   
```
missing end line

lpranam commented on test/core/histogram/access.cpp:
```diff
@@ -0,0 +1,35 @@
+    check_indexing_operator();
+
+    return boost::report_errors();
+}
```
missing end line

lpranam commented on test/core/histogram/is_compatible.cpp:
```diff
@@ -0,0 +1,72 @@
+    check_is_tuple_compatible();
+
+    return boost::report_errors();
+}
```
missing end line

lpranam commented on test/core/histogram/sub_histogram.cpp:
```diff
@@ -0,0 +1,58 @@
+    check_sub_histogram_with_tuple();
+
+    return boost::report_errors();
+}
```
missing end line

lpranam commented on test/core/histogram/key.cpp:
```diff
@@ -0,0 +1,62 @@
+    check_histogram_key_from_pixel();
+
+    return boost::report_errors();
+}   
```
missing end line

lpranam commented on test/core/histogram/helpers.cpp:
```diff
@@ -0,0 +1,73 @@
+    check_helper_fn_tuple_to_tuple();
+
+    return boost::report_errors();
+}   
```
missing end line

mloskot commented:
@codejaeger The subject of this PR includes words WIP and Test, does it mean the PR is still a draft?

Although any PR can be considered as WIP or draft until it is merged or rejected, I'd suggest to use tag WIP only to indicate as "Please, do not review it yet, I'm still working on it". Also, the "Test" is unnecessary and confusing.

If it is not a work in progress, I'd suggest to change the subject to "Add histogram class and basic operations" or similar.

codejaeger commented:
I had forgotten to remove the [WIP] tag. In the future I will resort to using the draft mode only to convey the state. 
The 'Test' can be removed now since the class idea is finalized now.

codejaeger commented:
@mloskot @lpranam Currently the histogram only supports integral bin types. I started with this for simplicity, but now I think floating point values are also needed for the algorithms. So, should we provide a histogramf wrapper that gives a floating point version of the histogram, so that the calling interface remains the same?

codejaeger commented:
@mloskot @lpranam I need to make some additions to the class which is needed for the histogram equalization algorithm. Should I add that to this PR or wait to add it with the algorithm?

lpranam commented on include/boost/gil/histogram.hpp:
```diff
@@ -428,11 +440,12 @@ histogram<T...> cumulative_histogram(histogram<T...>& hist)
     
     using histogram_t = histogram<T...>;
-    using pair_t = std::pair<typename histogram_t::key_type, typename histogram_t::mapped_type>;
+    using pair_t  = std::pair<typename histogram_t::key_type, typename histogram_t::mapped_type>;
```
nitpicking: remove the additional space before `=`

codejaeger replied:
I think it got added when I used clang-format, maybe to keep the equal signs in line 473 and 474 at the same position. This has been done in many other places also I think.

mloskot replied:
@codejaeger Which clang-format configuration file did you use?

codejaeger replied:
@mloskot I believe I had used the one linked below:
[https://github.com/boostorg/gil/blob/develop/example/clang-format/.clang-format](https://github.com/boostorg/gil/blob/develop/example/clang-format/.clang-format)

codejaeger commented on example/histogram.cpp:
```diff
@@ -1,49 +1,117 @@
 //
-// Copyright 2005-2007 Adobe Systems Incorporated
+// Copyright 2020 Debabrata Mandal <mandaldebabrata123@gmail.com>
```
The previous histogram.cpp got renamed to tutorial_histogram.cpp. I will add the copyright otherwise if needed.

mloskot replied:
Yes, I think I was suggesting that renaming and to sum it up
- `histogram.cpp` is @codejaeger 's new example
- `tutorial_histogram.cpp` is the old one

mloskot commented:
@codejaeger @lpranam First, I'm terribly sorry, for the long silence again.
I am catching up with the reviews now, but first I need to work out the merge of develop to master for Boost 1.74 release before the master closes today.

codejaeger commented:
@mloskot @lpranam I should make some patches to this PR, I can continue in this by marking it as draft. Is that fine?

mloskot commented:
@codejaeger I'm reviewing this PR, but please feel free to apply any changes you need.

@lpranam What's your take on this PR, I recall you had objections to some of the design choices. Please, forgive me that I had not reacted to that discussion in timely manner and that I can't remember exactly what was that about.

codejaeger commented:
> @codejaeger I'm reviewing this PR, but please feel free to apply any changes you need.
> 
> @codejaeger A bit of update to my comment:
> 
> The `example/histogram.cpp` is a bit too hard to follow. An example is supposed to be a minimum sample of code that presents how to use a feature.

@mloskot I agree it is difficult enough to read for anyone. The file started as an example and became more of a test in the later revisions. I will update it at once. I could also add block comments to separate out different sections of implementations (like create a histogram, fill it etc.). 

Although I don't expect anyone to come to this file for reference since I aim to make the docs self sufficient, I will make the changes required. Thanks for the suggestion.

mloskot commented:
@codejaeger 
> The file started as an example and became more of a test in the later revisions. 

Yes, that is quite common approach.

> Although I don't expect anyone to come to this file for reference since I aim to make the docs self sufficient, I will make the changes required. 

What you can do is to use the file to combine the snippets from the docs, so you can be sure they compile.

Ideally, if the examples in the docs are self-contained and are compile-tested when the docs are built, but I don't know how to achieve it in Sphinx. That is one of reasons why we may consider switch to AsciiDoc as some point in future

codejaeger commented:
@mloskot , @lpranam  I have 
1. updated the example file for histogram
2. reduced the warnings from the tests and example
3. Added Doxygen comments for the new changes
Unless the CI checks fail it is ready for review now.

mloskot commented:
@codejaeger Thanks for the update.

Something is wrong with the histogram example, see https://ci.appveyor.com/project/stefanseefeld/gil/builds/34407783/job/fw3u0ba7yno9rdui

The last build job of AppVeyor is CMake-based which builds examples too.

codejaeger commented:
Thanks @mloskot for pointing out the bug. I hope it gets fixed with the last commit.

mloskot reviewed (approved):
I don't have any major issues to report about this PR, so I'm happy to approve it.
The only 'unresolved' comments are about the `x<space><space>=<space>y` formatting due to use of the `clang-format` and pointed out by @lpranam which I'd consider not a stopper issue that can be addressed after the merge along with other updates or improvements that are applicable post-merge.

@codejaeger Once again, thanks for your contributions
